app/grandchallenge/reader_studies/forms.py

```diff
@@ -310,6 +337,52 @@ def interface_choices(self):
     def initial_interface(self):
         return self.interface_choices().first()
 
+    def widget_choices(self):
+        answer_type = self["answer_type"].value()
+        choices = [("", "Default")]
+        if answer_type:
+            try:
+                extra_options = ANSWER_TYPE_TO_QUESTION_WIDGET[answer_type]
+                choices.extend(
+                    (option.name, option.label) for option in extra_options
+                )
+            except KeyError:
+                pass
```

Comment:
We probably need an error here as the `ANSWER_TYPE_TO_QUESTION_WIDGET` would not have been updated by the programmer.

Reply:
I'm actually not sure if we want to throw an error here. We need to be able to return an emtpy list of options for a lot of answer types, and the easiest way to do this is to catch the KeyError here, right? The KeyError is also not very informative, a developer could forget to update the mapping and not get a KeyError if it happens to be for one of the already included answer types...

Reply:
The key error comes from looking up the answer type in `ANSWER_TYPE_TO_QUESTION_WIDGET`, so to return an empty list of options for a particular type you would define

```python
ANSWER_TYPE_TO_QUESTION_WIDGET = {
  ...
  NoWidgetAnswerType = []
}
```

If the `KeyError` is not informative then you transform it:

```python
try:
   foo["bar"]
except KeyError as e:
   raise MySpecialException("You need to update foo") from e
```

If a developer forgets to update the mapping, then yes, no way to catch that other than by testing.

Reply:
Yes, but if we always raise an error whenever the answer_type is not in the mapping dictionary, then a user will be unable to update the question in those cases.

Reply:
Yes, they shouldn't be able to as the application has not be programmed correctly.

Reply:
They should be able to update a question regardless of what the answer type is, they just shouldn't be able to define a widget for them. If we raise an error whenever there is a KeyError, they would not be able to change anything on the quesiton, the form would not load because the widget_choices() are evaluated every single time you load the question form.

Reply:
One of the foundational Python principles is that "Errors should never pass silently.". If you leave this as it is you open yourself up to bugs that are difficult to track down, as you are never aware of where the application is going wrong.

Yes, if the error is raised no one will be able to update the form. But, that is exactly the behaviour we want as the proper solution to this is a developer going back and updating `ANSWER_TYPE_TO_QUESTION_WIDGET` as we know it is out of date.

Reply:
I don't follow the logic here, sorry. The mapping is already out of date unless we add all answer types to it and if we add all answer types, there will never be a KeyError. Sorry, but I think I'm either missing something here, or I don't understand the principal.

Reply:
Oh, there will be a KeyError if we add a new answer_type. Is that what you mean?

Reply:
Yes, `ANSWER_TYPE_TO_QUESTION_WIDGET` should contain all answer types.